test/inductor/test_fused_attention.py

```diff
@@ -631,6 +677,10 @@ class SDPAPatternRewriterCudaTests(TestSDPAPatternRewriterTemplate):
         test_sdpa_rewriter_13_cuda = functools.partialmethod(
             TestSDPAPatternRewriterTemplate._test_sdpa_rewriter_13, dtype=torch.half
         )
+        test_sdpa_rewriter_14_cuda = functools.partialmethod(
+            TestSDPAPatternRewriterTemplate._test_sdpa_rewriter_14
+        )
+        # TODO test_sdpa_rewriter_15 for cuda failed pattern match, need to be enabled
 
```

Comment:
Do you have any more details on this ?

Reply:
Hi @eellison , here is the traceback msg for `test_sdpa_rewriter_15_cuda`. I am not sure why this error occurs.

```
ERROR: test_sdpa_rewriter_15_cuda (__main__.SDPAPatternRewriterCudaTests)
partial(func, *args, **keywords) - new function with partial application
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/testing/_internal/common_utils.py", line 1541, in wrapper
    return fn(*args, **kwargs)
  File "test/inductor/test_fused_attention.py", line 622, in _test_sdpa_rewriter_15
    self._check_common(dot_prod_attention, check_train=False)
  File "test/inductor/test_fused_attention.py", line 71, in _check_common
    result2, (source_code,) = run_and_get_code(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/utils.py", line 920, in run_and_get_code
    result = fn(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/eval_frame.py", line 417, in _fn
    return fn(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/eval_frame.py", line 580, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/convert_frame.py", line 384, in _convert_frame_assert
    return _compile(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/convert_frame.py", line 643, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/convert_frame.py", line 524, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/bytecode_transformation.py", line 1033, in transform_code_object
    transformations(instructions, code_options)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/convert_frame.py", line 151, in _fn
    return fn(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/convert_frame.py", line 489, in transform
    tracer.run()
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/symbolic_convert.py", line 2098, in run
    super().run()
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/symbolic_convert.py", line 780, in run
    and self.step()
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/symbolic_convert.py", line 743, in step
    getattr(self, inst.opname)(inst)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/symbolic_convert.py", line 2216, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/output_graph.py", line 914, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/home/eikan/local_disk/liaoxuan/anaconda3/envs/pt_gpu/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/output_graph.py", line 1085, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/output_graph.py", line 1157, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/output_graph.py", line 1138, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/__init__.py", line 1697, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/compile_fx.py", line 1177, in compile_fx
    return aot_autograd(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_functorch/aot_autograd.py", line 889, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_functorch/aot_autograd.py", line 602, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 427, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 632, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 97, in aot_dispatch_base
    compiled_fw = compiler(fw_module, updated_flat_args)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/compile_fx.py", line 1105, in fw_compiler_base
    return inner_compile(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/repro/after_aot.py", line 83, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/debug.py", line 304, in inner
    return fn(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/anaconda3/envs/pt_gpu/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/compile_fx.py", line 317, in compile_fx_inner
    compiled_graph = fx_codegen_and_compile(
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/compile_fx.py", line 551, in fx_codegen_and_compile
    compiled_fn = graph.compile_to_fn()
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/graph.py", line 1157, in compile_to_fn
    return self.compile_to_module().call
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/utils.py", line 911, in patched_compile_to_module
    mod = compile_to_module(self)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/graph.py", line 1105, in compile_to_module
    self.codegen_with_cpp_wrapper() if self.cpp_wrapper else self.codegen()
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/graph.py", line 1082, in codegen
    self.scheduler.codegen()
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/scheduler.py", line 2265, in codegen
    self.get_backend(device).codegen_nodes(node.get_nodes())
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/codegen/cuda_combined_scheduling.py", line 63, in codegen_nodes
    return self._triton_scheduling.codegen_nodes(nodes)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/codegen/triton.py", line 2680, in codegen_nodes
    return self.codegen_node_schedule(node_schedule, numel, rnumel)
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/codegen/triton.py", line 2836, in codegen_node_schedule
    src_code = kernel.codegen_kernel()
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/codegen/triton.py", line 2163, in codegen_kernel
    if self.gen_attr_descriptor_import():
  File "/home/eikan/local_disk/liaoxuan/pytorch/torch/_inductor/codegen/triton.py", line 2111, in gen_attr_descriptor_import
    import triton.compiler.compiler
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
ModuleNotFoundError: No module named 'triton.compiler.compiler'; 'triton.compiler' is not a package

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True


----------------------------------------------------------------------
Ran 1 test in 9.139s

FAILED (errors=1)
```

Reply:
This might just be a local failure in your build. Try unskipping and re-push?

Reply:
Oh, sorry... Here is the error msg reported in CI:

```
=================================== FAILURES ===================================
___________ SDPAPatternRewriterCudaTests.test_sdpa_rewriter_15_cuda ____________
Traceback (most recent call last):
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1541, in wrapper
    return fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/inductor/test_fused_attention.py", line 622, in _test_sdpa_rewriter_15
    self._check_common(dot_prod_attention, check_train=False)
  File "/var/lib/jenkins/workspace/test/inductor/test_fused_attention.py", line 76, in _check_common
    self.assertGreaterEqual(counters["inductor"]["fuse_attention"], 1)
  File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 1250, in assertGreaterEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 675, in fail
    raise self.failureException(msg)
AssertionError: 0 not greater than or equal to 1

To execute this test, run the following from the base repo dir:
     python test/inductor/test_fused_attention.py -k test_sdpa_rewriter_15_cuda
```

Reply:
@eellison The CUDA issue was resolved by adding param `device` in `torch.full`. Seems a bug of pattern matching.